[MM-17155] Deep Linking Landing Page

devinbinnie opened the pull request:
#### Summary
This PR contains a new landing page for deep linking. It is based on the work done here: https://github.com/mattermost/mattermost-webapp/pull/2299

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-17155

### Related Pull Requests
Server changes: https://github.com/mattermost/mattermost-server/pull/13362

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `0052f1c8c0ce32bc9d2838dd1f4663f97a492eb2`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `893259386d5776fe390a3a98a9c92d2646705f12`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `92bc67c7298253a8c8fddafdb2cdc5b020c6fe1d`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `95ebb2f8e96cd2dcde4918b46fca6c718b2a1022`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

jwilander commented on package.json:
```diff
@@ -14,6 +14,7 @@
     "compass-mixins": "0.12.10",
     "core-js": "3.3.6",
     "css-vars-ponyfill": "2.1.2",
+    "custom-protocol-detection": "github:ismailhabib/custom-protocol-detection",
```
Has this been checked out by a PM for licensing?

devinbinnie replied:
@esethna this was used in the original PR, are we okay here?

hmhealey replied:
It's MIT, so we should be fine there

https://github.com/ismailhabib/custom-protocol-detection

jwilander commented on components/linking_landing_page/index.tsx:
```diff
@@ -0,0 +1,28 @@
+function mapStateToProps(state: GlobalState) {
+    const config = getConfig(state);
+    const defaultTheme = getTheme(state);
+    const location = window.location.href.replace('/vault#', '');
```
I know this probably isn't a big deal since this component is meant to live at this URL but this replace is tightly coupling this index to the URL path it's being served at. If we were to move it to a different path this would break. Is there a way to do this replace differently?

Side note: Also feels wrong to be re-doing this replace every time mapStateToProps runs. Probably not a big deal since this page is likely very static.

devinbinnie replied:
Agreed on not running the `replace` in `mapStateToProps`.

I suppose we could have this be a config item, and the route name can be configured by the server if necessary. Does that work?

jwilander replied:
That could work but is pretty heavyweight. What about just grabbing the bits after the `#` and reconstructing the URL with siteURL pulled in from the config? Is that possible?

devinbinnie replied:
I suppose so, but that sounds more complicated than it needs to be. I've moved the `replace` function to `componentDidMount()` so that it's only run once now, which should make it less CPU-heavy. I think that should suffice, what do you think?

jwilander replied:
I'm good with that. Wasn't really worried too much about the CPU cost, more worried about the coupling to the URL but it's not a huge deal and it sounds like it's a pain to generalize. I'm good with it 👍

jwilander commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,354 @@
+
+    checkVaultPreference = () => {
+        const vaultPreference = BrowserStore.getVaultPreference(this.props.siteUrl);
+        if (vaultPreference) {
```
Minor but could reduce nesting a little by doing:

```
if (!vaultPreference) {
  return
}

switch (vaultPreferenence) {
...
```

jwilander commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,354 @@
+    }
+
+    setPreference = (pref: string) => {
+        if (this.state.rememberChecked) {
```
Could reduce nesting here as well

jwilander commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,354 @@
+        );
+    }
+
+    render() {
```
This is a pretty long render function, might be worth breaking up into smaller ones

jwilander commented on components/root/root.jsx:
```diff
@@ -300,12 +304,8 @@ export default class Root extends React.Component {
-                    <Route
-                        path={'/get_android_app'}
-                        component={GetAndroidApp}
+                        path={'/vault'}
```
Curious, why `/vault`?

devinbinnie replied:
It was the route name picked by the original PR. I'm not sure if there's any significance to it or not.

jwilander replied:
Can we go with something like `/landing` to be more clear?

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `45edf7c419f6cfa93c7ec340acbed03f5d3fffe7`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `f30560b2911525b36a59aae538e9135bc0c75151`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,392 @@
+}
+
+function protocolDetected() {
+    // Code in case everything worked
```
Dumb question, but is there supposed to be something here, or is it just handled by the native code at this point?

devinbinnie replied:
Oh yeah, this is leftover from the old PR. I'll remove it.

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,392 @@
+                <div className='get-app__dialog'>
+                    <div
+                        className={`get-app__graphic ${isMobile ? 'mobile' : ''}`}
+                        style={{backgroundColor: this.props.backgroundColor}}
```
Could we use CSS variables for this, or do those not exist at this point?

devinbinnie replied:
I believe we can actually, I'll look into that.

devinbinnie replied:
Done!

cpanato commented:
recreating the test server...

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `b63fac66d242e35a4212ab7cb3dfd63b5dab2340`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

esethna commented:
@devinbinnie I cannot seem to get to the landing page with the `/vault#` route. Can you please post instructions for testing here so design and QA can also test?

esethna commented:
Note to testers, access the landing page via "https://mattermost-webapp-pr-4455.test.mattermost.cloud/landing#"

@devinbinnie can you please write out what is testable in this PR for QA?

devinbinnie commented:
@esethna I've addressed your feedback, except for the custom logo bit, which I will need @andrewbrown00 's help on. I would imagine we would just resize the logo down to the same height as the provided logo such that it fits on the top bar, what do you think?

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

esethna commented:
@devinbinnie yes that sounds correct, with the site name to the right of it in similarly large font as in the "Where would you like to view this?" text. Also considering the case if there is no logo but there is site name

mattermod commented:
Mattermost test server updated with git commit `0ee24fb7390b12e40718382b90da75d6fbf9e958`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

esethna commented:
@devinbinnie can you review the expected in #2B and #2C from my review above?

devinbinnie commented:
> @devinbinnie can you review the expected in #2B and #2C from my review above?

@esethna  Misread that sorry, is there a reason we can't just sub in the site name for those strings?

esethna commented:
Subbing in the strings doesn't read as well imo. Plus it's possible that people have white labelled the webapp but not built their own custom desktop app (meaning it's still called Mattermost)

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `e3cb42584cdd5a37377b7887b1886f9cb9c43a2f`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

esethna commented:
@devinbinnie looks good other than the custom branding piece which I don't think was addressed. Can you ask for a re-review once everything is ready to go?

esethna commented:
@andrewbrown00 I think the black MM logo is fine. Also note that we can't use account setting theme variables because the user may not be logged in yet - so all we have access to is server configs which don't contain theme variables

esethna commented:
@andrewbrown00 please also test this on mobile safari to verify the mobile version. My iPhone X seems like it's missing some buffer room and the image is blurry:

![image](https://user-images.githubusercontent.com/14333569/71026658-8aa27880-20be-11ea-806b-e7094b748619.png)

andrewbrown00 commented:
> @andrewbrown00 please also test this on mobile safari to verify the mobile version. My iPhone X seems like it's missing some buffer room and the image is blurry:

@esethna interesting, I'm not seeing the spacing issues on my devices:

iPhone 6 Plus | iPhone 11 Pro Max
------------ | -------------
![iPhone-6-plus](https://user-images.githubusercontent.com/48559658/71040091-c86dc400-20f3-11ea-901e-0a44d130efa2.PNG) | ![iPhone-11-Pro-Max](https://user-images.githubusercontent.com/48559658/71040340-65306180-20f4-11ea-8051-f24b1157018e.png)

@devinbinnie any idea why the image is coming through blurry? Do I need to provide 2x and 3x versions of the image?

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

devinbinnie commented:
> @devinbinnie any idea why the image is coming through blurry? Do I need to provide 2x and 3x versions of the image?

I think it's probably due to the compression that webpack does on all the images. I also noticed it on the desktop image. We could try using a larger version for added clarity, is it possible to get a 3x version of the image?

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Mattermost test server updated with git commit `739b204e52ee1296bcc45058ab9338b13334ce53`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `f6d925e777966176a234068a2734dc22c0f39fe9`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

devinbinnie commented:
@andrewbrown00 All of your feedback should be addressed.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `716e588b3543801ad7834bfb664488c241028bf8`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

andrewbrown00 commented:
> > @devinbinnie any idea why the image is coming through blurry? Do I need to provide 2x and 3x versions of the image?
> 
> I think it's probably due to the compression that webpack does on all the images. I also noticed it on the desktop image. We could try using a larger version for added clarity, is it possible to get a 3x version of the image?

@devinbinnie thanks for the insight, here are the 3x versions. Let me know why they are up and I'll do another review.
[deep-linking-3x.zip](https://github.com/mattermost/mattermost-webapp/files/3980107/deep-linking-3x.zip)

devinbinnie commented:
@andrewbrown00 @esethna I've added the 3x images and it doesn't seem to be blurry anymore for me, let me know what you see.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `ba6bc290ea372faa5f5f88caee8e85886b13e7b6`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

andrewbrown00 commented:
> @andrewbrown00 @esethna I've added the 3x images and it doesn't seem to be blurry anymore for me, let me know what you see.

@devinbinnie Image is nice and crisp, thanks!

iPhone 6 Plus | iPhone 11 Pro Max
------------ | -------------
![IMG_E1E79266A807-1](https://user-images.githubusercontent.com/48559658/71138095-4f946800-21d8-11ea-8f62-12992dd7ec8a.jpeg) | ![IMG_5972](https://user-images.githubusercontent.com/48559658/71138022-12c87100-21d8-11ea-8ddd-021c3c688474.PNG)

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `d68d3987d398a855a9e6eb0af8a9099e243546d1`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `693befc03c8c2df33a86911c3250bd64e9f1359d`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

devinbinnie commented:
/update-branch

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `6e34070ab26fa4bcad680a3488c88cbd2307bf9c`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

devinbinnie commented:
/update-branch

mattermod commented:
Error trying to update the PR.
Please do it manually.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,473 @@
+            rememberChecked: false,
+            redirectPage: false,
+            location,
+            nativeLocation: location.replace(/^(https|http)/, 'mattermost'),
```
This doesn't really need to be in the state since it's easily derived from `this.state.location`. 0/5 on changing that now though

devinbinnie replied:
This was changed for @jwilander since he didn't want the `replace()` run more than once. So I added it in state.

hmhealey replied:
Ah, conflicting feedback. Well, I guess you can do it either way then 😛

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,473 @@
+        window.addEventListener('beforeunload', this.clearLandingPreferenceIfNotChecked);
+    }
+
+    clearLandingPreferenceIfNotChecked = () => {
```
Instead of clearing this preference when we navigate, why don't we just only save the value before navigation? It seems a little weird to save the value back to local storage and the clear it if the user doesn't actually click the button to navigate

devinbinnie replied:
We also save the value when we open the desktop app, which doesn't navigate. This gives us full consistent control over when the value is saved, even if it feels a little awkward.

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,473 @@
+        if (!this.state.navigating) {
+            BrowserStore.clearLandingPreference(this.props.siteUrl);
+        }
+        window.removeEventListener('beforeunload', this.clearLandingPreferenceIfNotChecked);
```
This should probably be removed in `componentWillUnmount` so that we don't leak anything if this component gets unmounted somehow without navigating. That'll mirror how other components do it anyway.

mattermod commented:
Mattermost test server updated with git commit `630758906b1e692d0919d9884b66616145081014`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,473 @@
+        this.setState({rememberChecked: !this.state.rememberChecked});
+    }
+
+    setPreference = (pref: string, clearIfNotChecked?: boolean) => {
```
Total nitpicking, but is there maybe some other name we could use for this instead of "preference" since that usually refers to the user preferences stored on the server?

devinbinnie replied:
Sure, but I can't think of one :P

hmhealey commented on components/linking_landing_page/linking_landing_page.tsx:
```diff
@@ -0,0 +1,473 @@
+        return (
+            <a
+                href={this.state.nativeLocation}
+                onMouseDown={() => {
```
Why mix `onMouseDown` and `onClick` here?

devinbinnie replied:
Because `onClick` will not fire if you middle click or Cmd+Click or right click to open in a new tab or window.

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `d98736921168b60166152a99e82068243ccbec18`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

devinbinnie commented:
/update-branch

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Mattermost test server updated with git commit `9aafd0f4f481dd8164ad99723e72ae9e9742c676`.

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

jgilliam17 commented:
@devinbinnie 
When I open the test server link in Chrome, I only see the Sign In page, same in incognito window. There's no "Where would you like to see this" option. Not sure how to get to it, I tried in Safari and Firefox and clearing cache and still don't see it.

In mobile:
`Testing 1234` for site name - is this just a placeholder text?
View in App - tap on the button View in App presents system message and "Opening link in the Desktop App..." below. Wording is confusing and it follows with "Don't have the Mobile App?" see screenshot.

When tapping `View in App` I expected system message to be overlaid on top of the landing page, so when I tap `Cancel` on the system message I would remain to the landing page, but it appears that redirect has already started. Is this intended? 
Thanks.

![IMG_079F2D6BA9F6-1](https://user-images.githubusercontent.com/52937121/72011463-ebd5eb80-3227-11ea-9815-285a5d4cccbc.jpeg)

andrewbrown00 commented:
> @devinbinnie
> When I open the test server link in Chrome, I only see the Sign In page, same in incognito window. There's no "Where would you like to see this" option. Not sure how to get to it, I tried in Safari and Firefox and clearing cache and still don't see it.
> 
> In mobile:
> `Testing 1234` for site name - is this just a placeholder text?
> View in App - tap on the button View in App presents system message and "Opening link in the Desktop App..." below. Wording is confusing and it follows with "Don't have the Mobile App?" see screenshot.
> 
> When tapping `View in App` I expected system message to be overlaid on top of the landing page, so when I tap `Cancel` on the system message I would remain to the landing page, but it appears that redirect has already started. Is this intended?
> Thanks.
> 
> ![IMG_079F2D6BA9F6-1](https://user-images.githubusercontent.com/52937121/72011463-ebd5eb80-3227-11ea-9815-285a5d4cccbc.jpeg)

mattermod commented:
Test server destroyed

andrewbrown00 commented:
Sorry, accidentally clicked close while reviewing 😬 My magic mouse has been freaking out since upgrading to macOS Catalina...

andrewbrown00 commented:
@devinbinnie will you create a new test server, see my previous comment, sorry...

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

devinbinnie commented:
> @devinbinnie
> When I open the test server link in Chrome, I only see the Sign In page, same in incognito window. There's no "Where would you like to see this" option. Not sure how to get to it, I tried in Safari and Firefox and clearing cache and still don't see it.

You'll need to be redirected from a specific link, like an email notification. You can use this link to test:
https://mattermost-webapp-pr-4455.test.mattermost.cloud/landing#/pr4455/channels/town-square

> In mobile:
> `Testing 1234` for site name - is this just a placeholder text?

Yes, when you define a site name and use custom branding, it will display on the landing page. Since I've refreshed the server, it should be reset back to Mattermost.

> View in App - tap on the button View in App presents system message and "Opening link in the Desktop App..." below. Wording is confusing and it follows with "Don't have the Mobile App?" see screenshot.

That's a bug, I'll fix that.
 
> When tapping `View in App` I expected system message to be overlaid on top of the landing page, so when I tap `Cancel` on the system message I would remain to the landing page, but it appears that redirect has already started. Is this intended?

No it isn't, I'll into that as well.

Thanks for the feedback :)

jwilander commented:
/update-branch

mattermod commented:
New commit detected. SpinWick will upgrade if the updated docker image is available.

mattermod commented:
Test server creation failed. See the logs for more information.

mattermod commented:
Test server destroyed

mattermod commented:
Creating a new SpinWick test server using Mattermost Cloud.

mattermod commented:
Mattermost test server created! :tada:

Access here: https://mattermost-webapp-pr-4455.test.mattermost.cloud

| Account Type | Username | Password |
|---|---|---|
| Admin | sysadmin | Sys@dmin123 |
| User | user-1 | User-1@123 |

mattermod commented:
Test server destroyed
